File: parsers/dart/test/value_test.dart
import 'package:synx/synx.dart';
import 'package:test/test.dart';

void main() {
  test('object set/get/remove', () {
    final o = SynxObject();
    o.set('a', synxInt(1));
    o.set('b', synxString('two'));
    expect(o['a'], equals(synxInt(1)));
    expect(o.contains('b'), isTrue);
    expect(o.remove('a'), isTrue);
    expect(o.contains('a'), isFalse);
  });

  test('object equality is order-insensitive', () {
    final a = SynxObject()
      ..set('x', synxInt(1))
      ..set('y', synxInt(2));
    final b = SynxObject()
      ..set('y', synxInt(2))
      ..set('x', synxInt(1));
    expect(synxObject(a), equals(synxObject(b)));
  });

  test('type helpers', () {
    expect(synxNull().isNull, isTrue);
    expect(synxInt(5).asInt, equals(5));
    expect(synxFloat(3.14).asDouble, equals(3.14));
    expect(synxString('x').asInt, isNull);
  });
}

File: parsers/go/value_test.go
package synx

import "testing"

func TestObjectSetGetRemove(t *testing.T) {
	o := NewObjectMap()
	o.Set("a", Int(1))
	o.Set("b", String("two"))
	if v, ok := o.Get("a"); !ok || !Equal(v, Int(1)) {
		t.Fatalf("get a failed")
	}
	if !o.Contains("b") {
		t.Fatalf("contains b failed")
	}
	if !o.Remove("a") {
		t.Fatalf("remove a failed")
	}
	if o.Contains("a") {
		t.Fatalf("a still present after remove")
	}
}

func TestObjectEqualityOrderInsensitive(t *testing.T) {
	a := NewObjectMap()
	a.Set("x", Int(1))
	a.Set("y", Int(2))
	b := NewObjectMap()
	b.Set("y", Int(2))
	b.Set("x", Int(1))
	if !Equal(Object_(a), Object_(b)) {
		t.Fatalf("order-insensitive equality failed")
	}
}

func TestTypeHelpers(t *testing.T) {
	if !IsNull(Null) {
		t.Fatalf("IsNull")
	}
	if n, ok := AsInt(Int(5)); !ok || n != 5 {
		t.Fatalf("AsInt")
	}
	if d, ok := AsNumber(Float(3.14)); !ok || d != 3.14 {
		t.Fatalf("AsNumber float")
	}
	if _, ok := AsInt(String("x")); ok {
		t.Fatalf("AsInt should fail on string")
	}
}
